axlearn/common/utils.py

```diff
@@ -1158,6 +1159,55 @@ def prune_tree(
     return in_tree
 
 
+def tree_merge_default_override_fn(primary: Any, secondary: Any):
+    is_primary_empty = False
+    is_secondary_empty = False
+    try:
+        is_primary_empty = len(primary) == 0
+        is_secondary_empty = len(secondary) == 0
+    except TypeError:
+        # A TypeError will be raised if primary/secondary don't have length,
+        # e.g. if they are scalars.
+        pass
+    if primary is None or is_primary_empty:
+        return secondary
+    if secondary is None or is_secondary_empty:
+        return primary
+    raise ValueError(
+        f"Encountered incompatible subtree leaves: {primary=}, {secondary=}. Specify "
+        "a custom override function to resolve incompatible subtree merges."
+    )
+
+
+def tree_merge(
+    primary: Nested[Any],
+    *,
+    secondary: Nested[Any],
+    override_fn: Callable[[Any, Any], Any] = tree_merge_default_override_fn,
+) -> Nested[Any]:
+    """Merge `secondary` into `primary`. The result contains shallow copies of subtrees from both.
+
+    Two trees are mergable if there does not exists a path in `secondary` that is a subpath of any
+    path in `primary`. If there are identical path with different leaves, `override_fn` is used to
+    determine which leaf is kept in the resulting tree.
+
+    The default `override_fn` choses the non-empty leaf. If both leaves are non-empty, an error
+    will be raised.
+    """
+    if isinstance(primary, dict) ^ isinstance(secondary, dict):
+        raise ValueError(f"Trying to merge incompatible subtrees: {primary=}, {secondary=}")
+    if not (isinstance(primary, dict) or isinstance(secondary, dict)):
+        return override_fn(primary, secondary)
+    # Use type() so that if primary is a VDict, out_tree is also a VDict.
```

Comment:
Can we be even more strict?
```suggestion
    if type(primary) != type(secondary):
        raise ValueError(...)
```

Reply:
Done.